marlowe-dashboard-client/src/Capability/PlutusApps/MarloweApp.purs

```diff
@@ -38,48 +47,134 @@ import Plutus.V1.Ledger.Value (TokenName) as Back
       backRoles = toBack roles
-    mutex <- asks $ view (_marloweAppEndpointMutex <<< _create)
-    liftAff $ AVar.take mutex
-    Contract.invokeEndpoint plutusAppId "create" (backRoles /\ contract)
+
+      payload = reqId Json./\ backRoles Json./\ contract
+    invokeMutexedEndpoint plutusAppId reqId "create" _create payload
   applyInputs plutusAppId marloweContractId (TransactionInput { interval: SlotInterval slotStart slotEnd, inputs }) = do
+    reqId <- liftEffect genUUID
     let
       backSlotInterval :: Back.Slot Json./\ Back.Slot
       backSlotInterval = (toBack slotStart) Json./\ (toBack slotEnd)
 
-      payload = marloweContractId Json./\ Just backSlotInterval Json./\ inputs
-    mutex <- asks $ view (_marloweAppEndpointMutex <<< _applyInputs)
-    liftAff $ AVar.take mutex
-    Contract.invokeEndpoint plutusAppId "apply-inputs" payload
+      payload = reqId Json./\ marloweContractId Json./\ Just backSlotInterval Json./\ inputs
+    invokeMutexedEndpoint plutusAppId reqId "apply-inputs" _applyInputs payload
   redeem plutusAppId marloweContractId tokenName pubKeyHash = do
+    reqId <- liftEffect genUUID
     let
-      payload :: MarloweParams Json./\ Back.TokenName Json./\ Back.PubKeyHash
-      payload = marloweContractId Json./\ toBack tokenName Json./\ toBack pubKeyHash
-    mutex <- asks $ view (_marloweAppEndpointMutex <<< _redeem)
-    -- TODO: we could later add a forkAff with a timer that unlocks this timer if we
-    --       dont get a response
-    liftAff $ AVar.take mutex
-    Contract.invokeEndpoint plutusAppId "redeem" payload
+      payload :: UUID Json./\ MarloweParams Json./\ Back.TokenName Json./\ Back.PubKeyHash
+      payload = reqId Json./\ marloweContractId Json./\ toBack tokenName Json./\ toBack pubKeyHash
+    invokeMutexedEndpoint plutusAppId reqId "redeem" _redeem payload
 
 createEndpointMutex :: Effect EndpointMutex
 createEndpointMutex = do
   create <- EAVar.empty
   applyInputs <- EAVar.empty
   redeem <- EAVar.empty
-  pure { create, applyInputs, redeem }
+  requests <- EAVar.new mempty
+  pure { create, applyInputs, redeem, requests }
+
+invokeMutexedEndpoint ::
+  forall payload m env.
+  Encode payload =>
+  MonadAff m =>
+  ManageContract m =>
+  MonadAsk (MarloweAppEndpointMutexEnv env) m =>
+  PlutusAppId ->
+  UUID ->
+  String ->
+  Lens' EndpointMutex (AVar Unit) ->
+  payload ->
+  m (AjaxResponse Unit)
+invokeMutexedEndpoint plutusAppId reqId endpointName _endpointMutex payload = do
+  -- There are three mutex involved in this operation:
+  --   * The endpointMutex help us avoid making multiple requests to the same endpoint if its not
+  --     available.
+  --   * The global requestMutex help us manage the request queue concurrently.
+  --   * For each request in the queue, we store one mutex that represents the response.
+  endpointMutex <- asks $ view (_marloweAppEndpointMutex <<< _endpointMutex)
+  requestMutex <- asks $ view (_marloweAppEndpointMutex <<< _requests)
```

Comment:
Can use `use`

```suggestion
  endpointMutex <- use $ _marloweAppEndpointMutex <<< _endpointMutex
  requestMutex <- use $ _marloweAppEndpointMutex <<< _requests
```

Reply:
good catch

Reply:
ah, actually I just tried it, but is not the same 😅. `use` uses the StateMonad, in here we are using the `Env` with the MonadAsk.

The reason to do it in the Env is that it would require the use of `Effect` or `Aff` to create the initial "state".

https://github.com/input-output-hk/plutus/blob/master/marlowe-dashboard-client/src/Env.purs#L20